OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -367,16 +398,18 @@ protected void BeginProduction(ProductionItem item)
 		// Returns the actor/trait that is most likely (but not necessarily guaranteed) to produce something in this queue
 		public virtual TraitPair<Production> MostLikelyProducer()
 		{
-			var trait = self.TraitsImplementing<Production>().FirstOrDefault(p => p.Info.Produces.Contains(Info.Type));
+			var trait = productionTraits.FirstOrDefault(p => !p.IsTraitDisabled && !p.IsTraitPaused && p.Info.Produces.Contains(Info.Type));
 			return new TraitPair<Production>(self, trait);
 		}
 
 		// Builds a unit from the actor that holds this queue (1 queue per building)
 		// Returns false if the unit can't be built
 		protected virtual bool BuildUnit(ActorInfo unit)
 		{
-			// Cannot produce if I'm dead
-			if (!self.IsInWorld || self.IsDead)
+			var enabledProductionTraits = productionTraits.Where(p => !p.IsTraitDisabled).ToArray();
```

Comment:
The uses of this below usually won't need to enumerate the whole collection, so the `ToArray()` will actually hurt performance.

Reply:
The best fix is to query `MostLikelyProducer` hereish, cancel the production if the trait is null, and then return false if it is paused or disabled.